diff --git a/html/janus.js b/html/janus.js
index 43f5e3ab..084046ac 100644
--- a/html/janus.js
+++ b/html/janus.js
@@ -1939,6 +1939,15 @@ function Janus(gatewayCallbacks) {
 					}
 				};
 				event.track.onunmute = function(ev) {
+					if (ev == null) {
+						// FIXME: Ugly hack for NULL event: https://github.com/pikvm/pikvm/issues/1084
+						try {
+							pluginHandle.onremotestream(null);
+							return;
+						} catch (e) {
+							Janus.error(e);
+						}
+					}
 					Janus.log("Remote track flowing again:", ev);
 					if(trackMutedTimeoutId != null) {
 						clearTimeout(trackMutedTimeoutId);

