From 8d9a289eefe9a409c24e550df18c29394e4fa2ef Mon Sep 17 00:00:00 2001
From: Maxim Devaev <mdevaev@gmail.com>
Date: Sat, 1 Nov 2025 23:06:56 +0200
Subject: [PATCH] pikvm: uac: fixed uninitialized set_audio()

---
 drivers/usb/gadget/function/u_audio.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/gadget/function/u_audio.c b/drivers/usb/gadget/function/u_audio.c
index ca8dbec65f73..8fca93f37e61 100644
--- a/drivers/usb/gadget/function/u_audio.c
+++ b/drivers/usb/gadget/function/u_audio.c
@@ -786,8 +786,11 @@ void u_audio_suspend(struct g_audio *audio_dev)
 {
 	struct snd_uac_chip *uac = audio_dev->uac;
 
-	set_active(&uac->p_prm, false);
-	set_active(&uac->c_prm, false);
+	// https://github.com/pikvm/pikvm/issues/1488
+	if (uac->p_prm.uac)
+		set_active(&uac->p_prm, false);
+	if (uac->c_prm.uac)
+		set_active(&uac->c_prm, false);
 }
 EXPORT_SYMBOL_GPL(u_audio_suspend);
 
-- 
2.51.2

